doc/specs/#476 - Pinned packages.md

```diff
@@ -0,0 +1,186 @@
+---
+author: Easton Pillay jedieaston/easton@planeteaston.com
+created on: 2022-01-25
+last updated: 2022-02-02
+issue id: 476
+---
+
+# Pinning a package
+
+For [#476](https://github.com/microsoft/winget-cli/issues/476)
+
+## Abstract
+
+This spec describes the functionality behind "pinning" a package (freezing a package at a certain version so that it will be not be automatically upgraded via `winget upgrade --all`).
+
+## Inspiration
+
+This functionality is inspired by functionality in other package managers, as well as community feedback. Many packages (software development tools and libraries commonly, but other software as well) introduce breaking changes that users may not want integrate into their workflow quite yet.
+
+## Solution Design
+
+A table of packages (by Name, Version and Product Code if available) that are currently pinned will be kept in the local tracking catalog. To be repository independent, these values should be correlated from local sources (Add and Remove Programs, Appx, etc). This also means that the user should be able to pin packages that were not installed via the Windows Package Manager, so that they won't be upgraded via `winget upgrade --all`. 
+
+
+## UI/UX Design
+
+Several commands will need to be added to support pinning packages. Where applicable, equivalent names (or new arguments) for PowerShell cmdlets will also be given.
+
+The main interface for the pinning feature will be a new command, `winget pin`, with a couple of options:
+
+
+```
+winget pin
+PowerShell: Get-WinGetPinnedPackage
+```
+
+Running `winget pin` with no arguments will show a table of currently pinned packages:
+
+```
+Name                                                 Id                                Version    
+------------------------------------------------------------------------------------------------
+Microsoft Bob                                        Microsoft.Bob                     2.35.0
+iTunes                                               Apple.iTunes                      12.0.199.4
+
+```
+
+```
+winget pin --set <package>
+PowerShell: Set-WinGetPinnedPackage
+```
+
+`<package>` correlates to a query, similar to `winget list`. When running this command, the package information will be stored in the table in the tracking catalog. This command should be able to accept Package Identifers (checked against all repositories) as well as names, tags, etc. If the package does not have a version number locally but is available from a source, then the version number from the source should be used.
+
+If there is no package directly available for software a user wants to pin, but it is [included as part of another package](https://github.com/microsoft/winget-cli/issues/1073), the package which includes said piece of software should be treated as pinned unless a newer version doesn't change the version of the pinned software. 
+
+```
+winget pin --clear <package>
+PowerShell: Remove-WinGetPinnedPackage
```

Comment:
I know this may be very opinion based, but to me, this powershell cmdlet naming makes me think it will uninstall the package that has been pinned. 

I would suggest any of the followng (and encourage discussion)-
`Lock-WinGetPackageVersion` and `Unlock-WinGetPackageVersion` - From [Common verbs](https://docs.microsoft.com/en-us/dotnet/api/System.Management.Automation.VerbsCommon?view=powershellsdk-1.1.0)
`Disable-WinGetPackageUpgrade` and `Enable-WinGetPackageUpgrade` - From [Lifecycle verbs](https://docs.microsoft.com/en-us/dotnet/api/System.Management.Automation.VerbsLifeCycle?view=powershellsdk-1.1.0)
`Suspend-WingetPackageUpgrade` and `Resume-WinGetPackageUpgrade` - From [Lifecycle verbs](https://docs.microsoft.com/en-us/dotnet/api/System.Management.Automation.VerbsLifeCycle?view=powershellsdk-1.1.0)